internal/tiltfile/flags/args_def.go

```diff
@@ -2,66 +2,155 @@ package flags
 
 import (
 	"bytes"
-	"errors"
+	"encoding/json"
 	"flag"
 	"fmt"
+	"os"
 
+	jsoniter "github.com/json-iterator/go"
+	"github.com/pkg/errors"
 	"go.starlark.net/starlark"
 
 	"github.com/windmilleng/tilt/internal/tiltfile/starkit"
+	"github.com/windmilleng/tilt/pkg/model"
 )
 
 type argValue interface {
-	flag() flag.Value
+	flag.Value
+	json.Marshaler
 	starlark() starlark.Value
 	setFromArgs([]string)
+	setFromInterface(interface{}) error
+	IsSet() bool
 }
 
 type argDef struct {
-	argValue
-	usage string
+	newValue func() argValue
+	usage    string
 }
 
 type ArgsDef struct {
 	positionalArgName string
 	args              map[string]argDef
 }
 
-func (ad ArgsDef) parse(args []string) (v starlark.Value, output string, err error) {
+func (ad ArgsDef) parse(flagsState model.FlagsState, args []string) (v starlark.Value, mergedArgs bool, output string, err error) {
```

Comment:
this function is very complex and difficult to follow. 

maybe try separating out the parts that do file i/o from the parts that don't do file i/o?

Reply:
alternatively, maybe this cries out for an ArgScanner struct that handles accumulated state, as described in:
https://dave.cheney.net/2019/01/27/eliminate-error-handling-by-eliminating-errors